# Chip Seq Peak Calling Workflow

Generated by BioPipelines Workflow Composer on 2025-11-25

## Description

Can you build chip-seq workflow to run?


## Analysis Type

chip_seq_peak_calling

## Organism

Not specified (genome build not specified)

## Modules Used

- fastqc: FASTQC
- bwamem: BWAMEM_ALIGN
- samtools: SAMTOOLS_SORT, SAMTOOLS_INDEX, SAMTOOLS_FLAGSTAT, SAMTOOLS_STATS, SAMTOOLS_IDXSTATS, SAMTOOLS_VIEW, SAMTOOLS_MERGE
- macs2: MACS2_CALLPEAK
- multiqc: MULTIQC, MULTIQC_CUSTOM
- picard: PICARD_MARKDUPLICATES, PICARD_COLLECTMULTIPLEMETRICS, PICARD_COLLECTRNASEQMETRICS, PICARD_COLLECTWGSMETRICS, PICARD_SORTSAM, PICARD_ADDORREPLACEREADGROUPS
- deeptools: DEEPTOOLS_BAMCOVERAGE, DEEPTOOLS_COMPUTEMATRIX, DEEPTOOLS_PLOTHEATMAP, DEEPTOOLS_PLOTPROFILE, DEEPTOOLS_MULTIBAMSUMMARY, DEEPTOOLS_PLOTCORRELATION, DEEPTOOLS_PLOTFINGERPRINT, DEEPTOOLS_PLOTPCA
- homer: HOMER_MAKETAGDIRECTORY, HOMER_FINDPEAKS, HOMER_ANNOTATEPEAKS, HOMER_FINDMOTIFSGENOME, HOMER_POS2BED, HOMER_MAKEUCSCFILE, HOMER_GETDIFFERENTIALPEAKS

## Usage

```bash
# Run with local executor
nextflow run main.nf -profile singularity

# Run on SLURM cluster
nextflow run main.nf -profile slurm,singularity

# Resume failed run
nextflow run main.nf -resume
```

## Input Files

Edit `samplesheet.csv` with your sample information.

## Output

Results will be saved to `results/` directory.

## Requirements

- Nextflow >= 23.04.0
- Singularity
- BioPipelines containers
