#!/usr/bin/env nextflow
nextflow.enable.dsl = 2

/*
 * Metagenomics Profiling Workflow
 * Generated by BioPipelines Workflow Composer
 * Date: 2025-11-25
 */

// Import modules
include { FASTQC } from './nextflow-modules/fastqc/main.nf'
include { KRAKEN2 } from './nextflow-modules/kraken2/main.nf'
include { METAPHLAN_PROFILE; METAPHLAN_MERGE } from './nextflow-modules/metaphlan/main.nf'
include { BRACKEN } from './nextflow-modules/bracken/main.nf'
include { MULTIQC; MULTIQC_CUSTOM } from './nextflow-modules/multiqc/main.nf'

// Parameters
params.reads = "data/raw/*_R{1,2}.fastq.gz"
params.kraken2_db = "data/databases/kraken2_standard"
params.metaphlan_db = "data/databases/metaphlan"
params.outdir = "results"

// Main workflow
workflow {
    // Input channels
    reads_ch = Channel
        .fromFilePairs(params.reads, checkIfExists: true)
        .map { sample_id, reads -> tuple([id: sample_id], reads) }
    
    // QC
    FASTQC(reads_ch)
    
    // Kraken2 classification
    KRAKEN2_CLASSIFY(reads_ch, file(params.kraken2_db))
    
    // Bracken abundance
    BRACKEN(KRAKEN2_CLASSIFY.out.report, file(params.kraken2_db), 150, 'S')
    
    // MetaPhlAn profiling
    METAPHLAN_PROFILE(reads_ch, file(params.metaphlan_db))
    
    // Merge profiles
    METAPHLAN_MERGE(METAPHLAN_PROFILE.out.profile.map { it[1] }.collect())
    
    // QC aggregation
    MULTIQC(FASTQC.out.zip.collect())
}
